src/rust/engine/process_execution/src/nailgun/nailgun_pool.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::io;
+use std::collections::HashMap;
+use std::io::BufRead;
+use std::path::PathBuf;
+use std::process::Stdio;
+use std::sync::Arc;
+
+use log::{debug, info, trace};
+use parking_lot::Mutex;
+use regex::Regex;
+
+use hashing::Digest;
+use lazy_static::lazy_static;
+
+use crate::ExecuteProcessRequest;
+
+lazy_static! {
+  static ref NAILGUN_PORT_REGEX: Regex = Regex::new(r".*\s+port\s+(\d+)\.$").unwrap();
+}
+
+pub type NailgunProcessName = String;
+type NailgunProcessFingerprint = Digest;
+type NailgunProcessMap = HashMap<NailgunProcessName, NailgunProcess>;
+pub type Port = usize;
+
+#[derive(Clone)]
+pub struct NailgunPool {
+    processes: Arc<Mutex<NailgunProcessMap>>,
+}
+
+impl NailgunPool {
+    pub fn new() -> Self {
+        NailgunPool {
+            processes: Arc::new(Mutex::new(NailgunProcessMap::new())),
+        }
+    }
+
+    ///
+    /// Given a `NailgunProcessName` and a `ExecuteProcessRequest` configuration,
+    /// return a port of a nailgun server running under that name and configuration.
+    ///
+    /// If the server is not running, or if it's running with a different configuration,
+    /// this code will start a new server as a side effect.
+    ///
+    pub fn connect(
```

Comment:
There is a lot of blocking code in here that should probably eventually be resolved by either 1) doing more of this on the Cpu/IO pool, 2) switching to tokio-process, as with the `LocalCommandRunner`. Not a "blocker" for this PR, but a potential deadlock risk to watch out for.

Reply:
I opened https://github.com/pantsbuild/pants/issues/8489, which will have to deal with this.